### ISSUE TITLE
Let Project & Settings inherit more shared methods

### ISSUE THREAD

**Issue by Omico** [gradle/gradle#27602]
### Expected Behavior

For now, we have to write like this

```kotlin
public class ExamplePlugin : org.gradle.api.Plugin<ExtensionAware> {
    override fun apply(target: ExtensionAware) {
        when (target) {
            is Project -> target.gradle.registerExampleServiceIfAbsent()
            is Settings -> target.gradle.registerExampleServiceIfAbsent()
        }
    }

    private fun Gradle.registerExampleServiceIfAbsent(): Provider<ExampleService> =
        sharedServices.registerIfAbsent(
            name = ExampleService.NAME,
            implementationType = ExampleService::class,
            configureAction = {},
        )
}
```

Provide an `GradleAware` interface, something like this
```java
// In org.gradle.api.Project
public interface Project extends Comparable<Project>, ExtensionAware, PluginAware, GradleAware

// In org.gradle.api.initialization.Settings
public interface Settings extends PluginAware, ExtensionAware, GradleAware

public interface GradleAware {
    Gradle getGradle();
}
```

Then it is more convenient to achieve this goal.

```kotlin
public class ExamplePlugin2 : org.gradle.api.Plugin<ExtensionAware> {
    override fun apply(target: ExtensionAware) {
        when (target) {
            is GradleAware -> target.gradle.registerExampleServiceIfAbsent()
        }
        when (target) {
            is Project -> // Do something for project
            is Settings ->  // Do something for settings
        } 
    }
}
```

If we go further, we can do something like this

```java
// TODO Maybe find a better name
public interface CommonAware extends PluginAware, ExtensionAware, GradleAware {
    BuildLayout getLayout();

    ProviderFactory getProviders();

    File getRootDir();
}

// In org.gradle.api.Project
public interface Project extends Comparable<Project>, CommonAware

// In org.gradle.api.initialization.Settings
public interface Settings extends CommonAware
```

### Context

It has been explained above.

**Comment by cobexer** [comment]
Thank you for your interest in Gradle!

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by mlopatkin** [comment]
In most cases, in particular, with `Gradle`, you can use `@Inject`. We do need to better document what is available for injection (this is tracked at #27427). For many other use cases, like `copy` or `exec`, there are dedicated services which are also injectable. Because of this, we are unlikely to invest more in this area.

Of course, please comment here or file a new issue if a new use case not covered by injection arises.

**Comment by Omico** [comment]
Hi @mlopatkin, I just facing a new case about reading a file from both `Settings` & `Project` from `rootDir`. I also need `ProviderFactory.gradleProperty()` to retrieve some value. 

It is a little dumb if I write something like this:

```
fun resolveXXX(rootDir: File, providers: ProviderFactory): XXX {
    rootDir.resolve("xxx")
    providers.gradleProperty("xxxxx").get()
    ...
}
```

I also need to call this function not just twice because my project has lots of composite build projects. It would be nice if we have the "CommonAware" like above. So that, I can write something like:
```
fun CommonAware.resolveXXX(): XXX {
    rootDir.resolve("xxx")
    providers.gradleProperty("xxxxx").get()
    ...
}
```
